### PR TITLE
Major - Split Point Identification + PostProcessor + Increased coverage

### PR DESCRIPTION
**Pull request by jerinphilip**
- Statistical split point Identification added.
- Rule based post-processing added.
- Wrote tests for trie, model and postprocessor classes.
- Added console scripts to train and test. Added pragma: no cover here.
- Reorganization of files
  - models are in sandhisplitter/models/
  - data in data/
  - informative debug logs in output/

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 travis:
 	python setup.py test --coverage \
-		--coverage-package-name=sandhisplitter
+		--coverage-package-name=sandhisplitter
```
Comment (copyninja) [inline Makefile]: Trailing whitespace?.. Clean it up

File: sandhisplitter/test.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import argparse
```
Comment (copyninja) [inline sandhisplitter/test.py]: Is this a command line tool?

File: sandhisplitter/tests/test_model.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+class TestModel(TestCase):
+    def setUp(self):
+        super(TestModel, self).setUp()
+        self.testModel = Model(depth=3, skip=1)
+
+    def test_load(self):
+        entries = [
+                u'വിഡ്‌ഢിയുണ്ടെങ്കില്‍=വിഡ്‌ഢി+ഉണ്ട്+എങ്കില്‍|7,11',
+                u'അകല്‍ച്ചയുടെസ്ഥാനത്തു്=അകല്‍ച്ചയുടെ+സ്ഥാനത്തു്|10',
```
Comment (copyninja) [inline sandhisplitter/tests/test_model.py]: Can we place test cases in a separate file under sandhisplitter/tests/resources and load it in setUp function?. If there are multiple tests we can utilize same entries in all tests

**Comment by copyninja** [comment]
I see multiple command line tools, are they for test or you intend to provide it?. If you intend to provide I would suggest keeping them in function instead **name** == '**main**' block and hook them to console_scripts in setup.cfg

**Comment by jerinphilip** [comment]
I'm not sure, I'll give you the purpose of the scripts and let you decide. If a developer wants to train using a different training data ( maybe in a different language ) these scripts will be helpful.

These scripts are to operate on model files. `train.py` generates the model after learning from annotated data and writes it to `models/`. `test.py` loads test data and checks the efficiency of the module. ( More features are to be implemented in `test.py`)

Once the above scripts serves it purpose and the models are saved and tested, the module needn't have them, or like you ship tests with the module you can ship the scripts too for a developer to hack. I suggest not to ship this as a command line tool since its most probably one time use.

File: sandhisplitter/tests/test_model.py (hunk 3 above)
Comment (jerinphilip) [inline sandhisplitter/tests/test_model.py]: Yeah, I think we should. These were select samples to get maximum coverage.
Now that I see`model.py` and `postprocessor.py` using similar testcases, its better to have them in one place. 

Will add this in the next PR.

Comment (jerinphilip) [inline sandhisplitter/tests/test_model.py]: Done.

File: Makefile (hunk 1 above)
Comment (jerinphilip) [inline Makefile]: Fixed.

**Comment by copyninja** [comment]
@jerinphilip Please rename test.py to something meaning ful. And both test.py and train.py should have their contents in some function and hook it up as console_scripts. From your explanation it looks like they  can be useful for developers so provide it as tools. For now I'm merging it do changes in a new PR.

**Comment by copyninja** [comment]
@jerinphilip Please rename test.py to something meaning ful. And both test.py and train.py should have their contents in some function and hook it up as console_scripts. From your explanation it looks like they  can be useful for developers so provide it as tools. For now I'm merging it do changes in a new PR.

File: ChangeLog [hunk 4]
```diff
@@ -1,6 +1,7 @@
 CHANGES
 =======
```
Comment (copyninja) [inline ChangeLog:2]: What do you mean by stash?.. Please provide meaningful comment. Rebase and update PR.

**Comment by copyninja** [comment]
There are many comments which does not make any sense. I think we have already told you to provide meaningful comments in commit. I will not merge this till you update commit message with meaningful descriptions. Use rebase to fix it.

**Comment by jerinphilip** [comment]
What do you want test.py renamed to? Its testing the model, therefore test is a meaningful name. How about benchmark.py?

Rebase done.

**Comment by copyninja** [comment]
> I'm not sure, I'll give you the purpose of the scripts and let you decide. If a developer wants to train using a different training data ( maybe in a different language ) these scripts will be helpful.

So you are saying these are tools that can be used to extend the splitter to new language

> These scripts are to operate on model files. train.py generates the model after learning from annotated data and writes it to models/. test.py loads test data and checks the efficiency of the module. ( More features are to be implemented in test.py)

train.py is definitely useful from your description. I still don't understand purpose of test.py for a developer. train.py and test.py are related or not?. Also you suggest more features can be implemented and below you again say we need not ship it.

> Once the above scripts serves it purpose and the models are saved and tested, the module needn't have them, or like you ship tests with the module you can ship the scripts too for a developer to hack. I suggest not to ship this as a command line tool since its most probably one time use.

Please make up your mind and please do not contradict your own statements. @stultus please let us know about your comments.

**Comment by jerinphilip** [comment]
There are two parameters, depth of the trie - which decides how many characters forward and backward needs to be taken into account and an initial skip, which affects accuracy. These are to be experimentally set to the optimum, based on [this](ltrc.iiit.ac.in/icon2015/icon2014_proceedings/papers/File71-p164.pdf) paper. 

Regarding train-test-application, please read [this](http://stats.stackexchange.com/a/19051).`test.py` outputs values of accuracy, precision etc which can be used to find out which model is better. More on the theory part - [wiki](https://en.wikipedia.org/wiki/Precision_and_recall)

It boils down to a choice, do you want these scripts in the packaged module ( `dist/` ) or not, which I'm leaving you the option to decide. I'd recommend keep these scripts outside `sandhisplitter`, and package the final module without the scripts. But thinking along the reasons you moved tests to `sandhisplitter/tests` has got me confused.

**Comment by aashiks** [comment]
From the conversations, I understand that this test.py is not unit tests, rather a useful tool (just like train.py). To make it clearer, sandhisplitter code does not depend on test.py or train.py, but are rather useful tools to generate and test the efficiency of models itself. 

Assuming my observations above, I recommend shipping them along with sandhisplitter as 'tools'. Rename test.py (testing the model is different from verifying that your software works properly, it is not meaningful in this contect) to something that reflects the fact that it verifies efficiency of model.

**Comment by copyninja** [comment]
@aashiks thanks for sharing your view. That is my idea too. So jerin hook it as console script which will be available for your user when he installs sandhisplitter.

Next up please write up some documentation code wise as well as using these tools.

**Comment by copyninja** [comment]
While rebasing and squashing commits remove unwanted commit messages. And commit atomically, i.e. only related changes should go in one commit. An example following commit

> Added tox for automated testing, Split point identification working.
> 
> Model separated, better suited for testing. Added tox
> 
> Stash
> 
> Improvements
> 
> Major changes, split point identification working

It looks like its a squash of some 5 commits which are totally unrelated. Improvements and Majour changes to split point identification could have gone together. I don't know what stash means and tox related could have gone together.

Idea is allow at any point of time go back and take one particular commit we are interested. Here if I want to debug your split point work only I've no way to use that because it also brings some other unwanted changes.

I don't think it can be fixed now so please keep this in mind and do atomic commits.

**Pull request merged**